File: image-processing/python/image-processing.py
import requests
from io import BytesIO
from PIL import Image, ImageFilter

image_url = "https://raw.githubusercontent.com/mikolalysenko/lena/master/lena.png"


def apply_filters(image_url, filter):
    if not filter:
        print("Filter is not provided.")
        return
    response = requests.get(image_url)

    if response.status_code == 200:
        img = Image.open(BytesIO(response.content))
    else:
        print(f"Failed to fetch image: {response.status_code}")
        print(response.text[:500])
    filtered_img = None
    # Apply filter
    if filter == "blur":
        blurred_img = img.filter(ImageFilter.GaussianBlur(radius=5))
        filtered_img = blurred_img
    elif filter == "grayscale":
        grayscale_img = img.convert("L")
        filtered_img = grayscale_img
    elif filter == "unsharp":
        unsharp_img = img.filter(ImageFilter.UnsharpMask(radius=5))
        filtered_img = unsharp_img
    else:
        print("Unknown filter.")
        return

    return filtered_img

if __name__ == "__main__":
    filtered_img = apply_filters(image_url=image_url, filter="unsharp")
    filename = "/data/outputs/filtered_image.png"
    filtered_img.save(filename)
    print(f"Filters applied and images saved successfully as {filename}")

File: image-processing/node/image-processing.js
const axios = require("axios");
const sharp = require("sharp");
const fs = require("fs");
const path = require("path");

const imageUrl =
  "https://raw.githubusercontent.com/mikolalysenko/lena/master/lena.png";

async function applyFilters(imageUrl, filter) {
  if (!filter) {
    console.log("Filter is not provided.");
    return;
  }

  const response = await axios.get(imageUrl, { responseType: "arraybuffer" });
  const imageBuffer = Buffer.from(response.data);

  let processedImage;

  switch (filter) {
    case "blur":
      processedImage = await sharp(imageBuffer).blur(5).toBuffer();
      break;
    case "grayscale":
      processedImage = await sharp(imageBuffer).grayscale().toBuffer();
      break;
    case "unsharp":
      processedImage = await sharp(imageBuffer).sharpen(5, 1, 1).toBuffer();
      break;
    default:
      console.log("Unknown filter.");
      return;
  }

  return processedImage;
}

async function main() {
  const filteredImage = await applyFilters(imageUrl, "grayscale");

  const outputDir = path.join("data", "outputs");
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const outputPath = path.join(outputDir, "filtered_image.png");
  await fs.promises.writeFile(outputPath, filteredImage);

  console.log(`Filters applied and image saved successfully as ${outputPath}`);
}

main();
